go/lib.h

```diff
@@ -0,0 +1,99 @@
+ *
+ * `connection_request_bytes` is an array of bytes that will be parsed into a Protobuf `ConnectionRequest` object.
+ * `connection_request_len` is the number of bytes in `connection_request_bytes`.
+ * `success_callback` is the callback that will be called in the case that the Redis command succeeds.
+ * `failure_callback` is the callback that will be called in the case that the Redis command fails.
+ *
+ * # Safety
+ *
+ * * `connection_request_bytes` must point to `connection_request_len` consecutive properly initialized bytes. It should be a well-formed Protobuf `ConnectionRequest` object. The array must be allocated on the Golang side and subsequently freed there too after this function returns.
+ * * `connection_request_len` must not be greater than the length of the connection request bytes array. It must also not be greater than the max value of a signed pointer-sized integer.
+ * * The `conn_ptr` pointer in the returned `ConnectionResponse` must live until it is passed into `close_client`.
+ * * The `connection_error_message` pointer in the returned `ConnectionResponse` must live until the returned `ConnectionResponse` pointer is passed to `free_connection_response`.
+ * * Both the `success_callback` and `failure_callback` function pointers need to live until the client is closed via `close_client` since they are used when issuing Redis commands.
+ */
+const struct ConnectionResponse *create_client(const uint8_t *connection_request_bytes,
+                                               uintptr_t connection_request_len,
+                                               SuccessCallback success_callback,
+                                               FailureCallback failure_callback);
+
+/**
+ * Closes the given client, deallocating it from the heap.
+ *
+ * `client_ptr` is a pointer to the client returned in the `ConnectionResponse` from `create_client`.
+ *
+ * # Safety
+ *
+ * * `client_ptr` must be obtained from the `ConnectionResponse` returned from `create_client`.
+ * * `client_ptr` must be valid until `close_client` is called.
+ * * `client_ptr` must not be null.
+ */
+void close_client(const void *client_ptr);
+
+/**
+ * Deallocates a `ConnectionResponse`.
+ *
+ * This function also frees the contained error.
+ *
+ * # Safety
+ *
+ * * `connection_response_ptr` must be obtained from the `ConnectionResponse` returned from `create_client`.
+ * * `connection_response_ptr` must be valid until `free_connection_response` is called.
+ * * `connection_response_ptr` must not be null.
+ * * The contained `connection_error_message` must be obtained from the `ConnectionResponse` returned from `create_client`.
+ * * The contained `connection_error_message` must be valid until `free_connection_response` is called and it must outlive the `ConnectionResponse` that contains it.
+ * * The contained `connection_error_message` must not be null.
```

Comment:
isn't connection_error_message null if an error did not occur while connecting though?

Reply:
This is outdated from when we weren't null-checking the error message but had a separate method to free it. I'll remove this.